homeassistant/components/onkyo/media_player.py

```diff
@@ -106,462 +203,405 @@
+
+        setup_receiver(receiver)
     else:
-        for receiver in eISCP.discover():
+
+        @callback
+        async def async_onkyo_discovery_callback(receiver):
+            """Receiver discovered, connection not yet active."""
+            _LOGGER.debug("Receiver discovered: %s (%s)", receiver.name, receiver.host)
             if receiver.host not in KNOWN_HOSTS:
-                hosts.append(OnkyoDevice(receiver, config.get(CONF_SOURCES)))
-                KNOWN_HOSTS.append(receiver.host)
-    add_entities(hosts, True)
+                await receiver.connect()
+                setup_receiver(receiver)
 
+        await pyeiscp.Connection.discover(
+            update_callback=async_onkyo_update_callback,
+            connect_callback=async_onkyo_connect_callback,
+            discovery_callback=async_onkyo_discovery_callback,
+        )
+
+    @callback
+    def close_receiver(_event):
+        for receiver in receivers.values():
+            receiver.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close_receiver)
 
-class OnkyoDevice(MediaPlayerEntity):
-    """Representation of an Onkyo device."""
 
-    _attr_supported_features = SUPPORT_ONKYO
+class OnkyoMediaPlayer(MediaPlayerEntity):
+    """Representation of an Onkyo Receiver Media Player (one per each zone)."""
+
+    _attr_should_poll = False
+    _attr_sound_mode_list = list(SOUND_MODE_MAPPING)
 
     def __init__(
-        self,
-        receiver,
-        sources,
-        name=None,
-        max_volume=SUPPORTED_MAX_VOLUME,
-        receiver_max_volume=DEFAULT_RECEIVER_MAX_VOLUME,
-    ):
+        self, receiver, sources, zone, max_volume, receiver_max_volume
+    ) -> None:
         """Initialize the Onkyo Receiver."""
+        super().__init__()
         self._receiver = receiver
-        self._attr_is_volume_muted = False
-        self._attr_volume_level = 0
-        self._attr_state = MediaPlayerState.OFF
-        if name:
-            # not discovered
-            self._attr_name = name
-        else:
+        name = receiver.name
+        self._attr_name = f"{name}{' ' + ZONES[zone] if zone != 'main' else ''}"
+        identifier = receiver.identifier
+        if identifier is not None:
             # discovered
-            self._attr_unique_id = (
-                f"{receiver.info['model_name']}_{receiver.info['identifier']}"
-            )
-            self._attr_name = self._attr_unique_id
+            self._attr_unique_id = f"{identifier}_{zone}"
+        else:
+            # not discovered
+            self._attr_unique_id = None
 
+        self._zone = zone
+        self._source_mapping = sources
+        self._reverse_mapping = {value: key for key, value in sources.items()}
         self._max_volume = max_volume
         self._receiver_max_volume = receiver_max_volume
+
+        self._attr_state = None
+        self._attr_is_volume_muted = False
+        self._attr_volume_level = 0
+        self._attr_source = None
```

Comment:
Can be set outside of the constructor

Reply:
I'm not sure I like setting instance variables initial values as class variables. Is that typical in the code base here?
I would prefer leaving as is.

Reply:
If we set them to non mutable values we almost always initialize them outside of the constructor